ocaml/alerts/expiry_alert.ml

```diff
@@ -0,0 +1,168 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module XenAPI = Client.Client
+module Date = Xapi_stdext_date.Date
+
+type message_name_t = string
+
+type message_priority_t = int64
+
+type message_id_t = message_name_t * message_priority_t
+
+type remaining_days_t = int
+
+type expiry_messaging_info_t = {
+    message_cls:
+      [ `Certificate
+      | `Host
+      | `PVS_proxy
+      | `Pool
+      | `SR
+      | `VDI
+      | `VM
+      | `VMPP
+      | `VMSS ]
+  ; message_obj_uuid: string
+  ; obj_description: string
+  ; message_sent_on_remaining_days_list: (remaining_days_t * message_id_t) list
+  ; expiry: Xapi_stdext_date.Date.t (* when the obj will expire *)
+}
+
+let seconds_per_day = 3600. *. 24.
+
+let days_until_expiry epoch expiry =
+  (expiry /. seconds_per_day) -. (epoch /. seconds_per_day)
+
+let all_messages rpc session_id =
+  XenAPI.Message.get_all_records ~rpc ~session_id
+
+let related_messages msg_obj_uuid related_message_name_list all_msgs =
+  List.filter
+    (fun (_ref, record) ->
+      let related name = List.mem name related_message_name_list in
+      record.API.message_obj_uuid = msg_obj_uuid
+      && related record.API.message_name
+    )
+    all_msgs
+
+let message_body msg expiry =
+  let valid_till = Date.of_float (Date.to_float expiry -. 1.) in
```

Comment:
It is because the `expiry` time in the `Expiry_alert` module is taken as the first second when something will expire, the last second when it has not expired is 1 second before `expiry`.
While I think what we need to print in the message body in the `date` field for alerting is the last second when it is still valid. So I add back 1 second for it.

Reply:
Changed to avoid the "-1 and +1" in [5da53be](https://github.com/xapi-project/xen-api/pull/5060/commits/5da53be8c5a09d72ea915f758fe14efdad404edf)